homeassistant/components/sensor/season.py

```diff
@@ -0,0 +1,149 @@
+"""
+Support for tracking which astronomical or meteorological season it is.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/sensor/season/
+"""
+import logging
+from datetime import datetime
+
+import voluptuous as vol
+
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import STATE_UNKNOWN, CONF_TYPE
+from homeassistant.helpers.entity import Entity
+import homeassistant.util as util
+
+REQUIREMENTS = ['ephem==3.7.6.0']
+
+_LOGGER = logging.getLogger(__name__)
+
+NORTHERN = 'northern'
+SOUTHERN = 'southern'
+EQUATOR = 'equator'
+STATE_SPRING = 'Spring'
+STATE_SUMMER = 'Summer'
+STATE_AUTUMN = 'Autumn'
+STATE_WINTER = 'Winter'
+TYPE_ASTRONOMICAL = 'astronomical'
+TYPE_METEOROLOGICAL = 'meteorological'
+VALID_TYPES = [TYPE_ASTRONOMICAL, TYPE_METEOROLOGICAL]
+
+HEMISPHERE_SEASON_SWAP = {STATE_WINTER: STATE_SUMMER,
+                          STATE_SPRING: STATE_AUTUMN,
+                          STATE_AUTUMN: STATE_SPRING,
+                          STATE_SUMMER: STATE_WINTER}
+
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_TYPE, default=TYPE_ASTRONOMICAL): vol.In(VALID_TYPES)
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Display the current season."""
+    if None in (hass.config.latitude, hass.config.longitude):
+        _LOGGER.error("Latitude or longitude not set in Home Assistant config")
+        return False
+
+    latitude = util.convert(hass.config.latitude, float)
+    _type = config.get(CONF_TYPE)
+
+    if latitude < 0:
+        hemisphere = SOUTHERN
+    elif latitude > 0:
+        hemisphere = NORTHERN
+    else:
+        hemisphere = EQUATOR
+
+    _LOGGER.debug(_type)
+    add_devices([Season(hass, hemisphere, _type)])
+
+    return True
+
+
+class Season(Entity):
+    """Representation of the current season."""
+
+    def __init__(self, hass, hemisphere, season_tracking_type):
+        """Initialize the season."""
+        self.hass = hass
+        self.hemisphere = hemisphere
+        self.datetime = datetime.now()
+        self.type = season_tracking_type
+        self.season = self.get_season()
+
+    @property
+    def name(self):
+        """Return the name."""
+        return "Season"
+
+    @property
+    def state(self):
+        """Return the current season."""
+        return self.season
+
+    def get_season(self):
```

Comment:
I think that a nice cleanup would be to extract this method from this class and just make it a method in this file . Have it pass datetime as a parameter.

That way it will make your tests also a lot easier, you can test the method directly.